src/image/ExposureFitsReader.cc

```diff
@@ -0,0 +1,358 @@
+        std::shared_ptr<daf::base::PropertyList> imageMetadata
+    ) {
+        // Try to read WCS from image metadata, and if found, strip the keywords used
+        try {
+            wcs = afw::geom::makeSkyWcs(*imageMetadata, true);
+        } catch (lsst::pex::exceptions::TypeError) {
+            LOGLS_DEBUG(_log, "No WCS found in FITS metadata");
+        }
+
+        // Strip LTV1, LTV2 from imageMetadata, because we don't use it internally
+        imageMetadata->remove("LTV1");
+        imageMetadata->remove("LTV2");
+
+        if (!imageMetadata->exists("INHERIT")) {
+            // New-style exposures put everything but the Wcs in the primary HDU, use
+            // INHERIT keyword in the others.  For backwards compatibility, if we don't
+            // find the INHERIT keyword, we ignore the primary HDU metadata and expect
+            // everything to be in the image HDU metadata.  Note that we can't merge them,
+            // because they're probably duplicates.
+            metadata = imageMetadata;
+        } else {
+            metadata = primaryMetadata;
+        }
+
+        filter = Filter(metadata, true);
+        detail::stripFilterKeywords(metadata);
+
+        visitInfo = std::make_shared<VisitInfo>(*metadata);
+        detail::stripVisitInfoKeywords(*metadata);
+
+        calib = std::make_shared<Calib>(metadata);
+        detail::stripCalibKeywords(metadata);
+    }
+
+    std::shared_ptr<daf::base::PropertyList> metadata;
+    Filter filter;
+    std::shared_ptr<afw::geom::SkyWcs> wcs;
+    std::shared_ptr<Calib> calib;
+    std::shared_ptr<VisitInfo> visitInfo;
+};
+
+
+class ExposureFitsReader::ArchiveReader{
+public:
+
+    enum Component {
+        PSF=0,
+        WCS,
+        COADD_INPUTS,
+        AP_CORR_MAP,
+        VALID_POLYGON,
+        TRANSMISSION_CURVE,
+        N_ARCHIVE_COMPONENTS
+    };
+
+    explicit ArchiveReader(daf::base::PropertyList & metadata) {
+        auto popInt = [&metadata](std::string const& name) {
+            int r = 0;
+            if (metadata.exists(name)) {
+                r = metadata.get<int>(name);
+                metadata.remove(name);
+            }
+            return r;
+        };
+        _hdu = popInt("AR_HDU");
+        _ids[PSF] = popInt("PSF_ID");
+        _ids[WCS] = popInt("SKYWCS_ID");
+        _ids[COADD_INPUTS] = popInt("COADD_INPUTS_ID");
+        _ids[AP_CORR_MAP] = popInt("AP_CORR_MAP_ID");
+        _ids[VALID_POLYGON] = popInt("VALID_POLYGON_ID");
+        _ids[TRANSMISSION_CURVE] = popInt("TRANSMISSION_CURVE_ID");
+    }
+
+    template <typename T>
+    std::shared_ptr<T> readComponent(afw::fits::Fits * fitsFile, Component c) {
+        if (!_ensureLoaded(fitsFile)) {
+            return nullptr;
+        }
+        return _archive.get<T>(_ids[c]);
+    }
+
+private:
+
+    bool _ensureLoaded(afw::fits::Fits * fitsFile) {
+        if (!_isLoaded && _hdu) {
+            --_hdu;  // archive ID in header is 1-indexed
```

Comment:
I suppose it'd be best to move the decrement to immediately after we read the header value (HDU indices in FITS are 1-indexed, but our internal code is zero-indexed), and use a separate variable rather than a sentinal value to indicate when the archive is not present.

Reply:
Are you sure?  We used to have a discrepancy, but it was based on a sloppy reading by me of the cfitsio (I think) standard.  There'll be a discussion on a ticket, but I forget which one.